src/tsobj-file-system-provider.ts

```diff
@@ -0,0 +1,264 @@
+import * as vscode from 'vscode';
+import * as path from 'path';
+import { exec } from 'child_process';
+import { Tailscale } from './tailscale/cli';
+import { Logger } from './logger';
+
+export class File implements vscode.FileStat {
+  type: vscode.FileType;
+  ctime: number;
+  mtime: number;
+  size: number;
+
+  name: string;
+  data?: Uint8Array;
+
+  constructor(name: string) {
+    this.type = vscode.FileType.File;
+    this.ctime = Date.now();
+    this.mtime = Date.now();
+    this.size = 0;
+    this.name = name;
+  }
+}
+
+export class Directory implements vscode.FileStat {
+  type: vscode.FileType;
+  ctime: number;
+  mtime: number;
+  size: number;
+
+  name: string;
+  entries: Map<string, File | Directory>;
+
+  constructor(name: string) {
+    this.type = vscode.FileType.Directory;
+    this.ctime = Date.now();
+    this.mtime = Date.now();
+    this.size = 0;
+    this.name = name;
+    this.entries = new Map();
+  }
+}
+
+export type Entry = File | Directory;
+
+export class TSObjFileSystemProvider implements vscode.FileSystemProvider {
+  // Implementation of the `onDidChangeFile` event
+  onDidChangeFile: vscode.Event<vscode.FileChangeEvent[]> = new vscode.EventEmitter<
+    vscode.FileChangeEvent[]
+  >().event;
+
+  watch(): vscode.Disposable {
+    throw new Error('Watch not supported');
+  }
+
+  stat(uri: vscode.Uri): vscode.FileStat | Thenable<vscode.FileStat> {
+    Logger.info(`stat: ${uri.toString()}`, 'tsobj-fsp');
+    const { hostname, resourcePath } = this.extractHostAndPath(uri);
+
+    const command = `ssh ${hostname} "stat -L -c '{\\"type\\": \\"%F\\", \\"size\\": %s, \\"ctime\\": %Z, \\"mtime\\": %Y}' ${resourcePath}"`;
+
+    return new Promise((resolve, reject) => {
+      exec(command, (error, stdout) => {
+        if (error) {
+          reject(error);
+        } else {
+          const result = JSON.parse(stdout.trim());
+          const type =
+            result.type === 'directory' ? vscode.FileType.Directory : vscode.FileType.File;
+          const size = result.size || 0;
+          const ctime = result.ctime * 1000;
+          const mtime = result.mtime * 1000;
+          resolve({ type, size, ctime, mtime });
+        }
+      });
+    });
+  }
+
+  readDirectory(uri: vscode.Uri): Promise<[string, vscode.FileType][]> {
+    Logger.info(`readDirectory: ${uri.toString()}`, 'tsobj-fsp');
+
+    const { hostname, resourcePath } = this.extractHostAndPath(uri);
+    Logger.info(`hostname: ${hostname}`, 'tsobj-fsp');
+    Logger.info(`remotePath: ${resourcePath}`, 'tsobj-fsp');
+
+    const command = `ssh ${hostname} ls -lA "${resourcePath}"`;
```

Comment:
can we just run `ls -ap` here? it wouldn't require skipping lines or anything.

Reply:
Yeah, we can do `ls -Ap`. `A` will exclude the `.` and `..` special entries.